readme.md
# @brixtol/cryptographer

Cryptographer with IV used for encryption and decryption of various data types with [crypto](https://nodejs.org/api/crypto.html). Uses an [aes-256-gcm](https://en.wikipedia.org/wiki/Galois/Counter_Mode) algorithm and supports multiple ciphers.

## Why?

For encryption/decryption of sensitive data sent across the wire and processed within serverless functions. We opened sourced the package which is merely a wrapper for encryption/decryption with Crypto.

## Install

```cli
pnpm add @brixtol/cryptographer
```

## Usage

```ts
import { Cryptographer, md5 } from "@brixtol/cryptographer";

const crypto = Cryptographer(
  secret: "secret"
  , algorithm?: "aes-256-ctr"
  , options?: CipherCCMOptions
);

// Encoding
crypto.encode({ foo: "bar" });

// Decoding, eg: { foo: "bar" }
crypto.decode("12345678910abcdefghijkmnopqrstuvwxyz");

// Create a md5 hash
md5('hello world') // => 5eb63bbbe01eeed093cb22bb8f5acdc3

```

## Ciphers

- aes-256-cbc
- aes-256-cbc-hmac-sha1
- aes-256-cbc-hmac-sha256
- aes-256-cfb
- aes-256-cfb1
- aes-256-cfb8
- aes-256-ctr
- aes-256-ofb
- aes256
- camellia256

### License

Licensed under [MIT](#LICENCE)

---

We [♡](https://www.brixtoltextiles.com/discount/4D3V3L0P3RS]) open source!


changelog.md
## _CHANGELOG_

## v0.3.1 - 04/10/2021

- Fix declaration generation
- Update rollup config to work with @rollup/typescript
  - Note to self: _Rollup typescript plugins are absolute fucking trash_


## v0.3.0 - 04/10/2021

- Added md5 utility helper
- Updated rollup typescript plugin

## v0.2.0 - 03/09/2021

- Converted Project to typescript
- Added monorepo deps


test/index.test.js
// @ts-nocheck

import test from 'ava';
import { Cryptographer, md5 } from '../package/index.cjs';

test.before((t) => {
  t.context.crypto = Cryptographer('$ecret');
});

test.serial('Create md5 string', (t) => {

  const value = 'hello world';
  const hash = md5(value);

  t.not(hash, value);

  t.log('md5 before: ', value);
  t.log('md5 after: ', hash);

  t.pass('md5 text passes');

});

test.serial('String value to encode and decode', (t) => {

  const encode = t.context.crypto.encode('some string');
  const decode = t.context.crypto.decode(encode);

  t.is('some string', decode);
  t.pass('String test passes');

});

test.serial('Incorrect secret', (t) => {

  const crypto = Cryptographer('invalid iv');

  // use the above instance to decode
  const encode = crypto.encode('want to fight?');

  // use context instance
  const decode = t.context.crypto.decode(encode);

  t.not('want to fight?', decode);
  t.pass('Object test passes');

});

test.serial('Object value to encode and decode', (t) => {

  const encode = t.context.crypto.encode({ foo: 'bar', bar: 'foo' });
  const decode = t.context.crypto.decode(encode);

  t.deepEqual({ foo: 'bar', bar: 'foo' }, decode);
  t.pass('Object test passes');

});


src/index.js
import { createHash, createDecipheriv, createCipheriv, getCiphers, getHashes } from 'crypto'
import os from 'os'

/**
 * Creates hash of an string based on available hashes of platform
 *
 * @param {string} input
 * @param {string} hash
 * @returns {string}
 */
const hash = (input, hash = 'md5') => {

  if (getHashes().indexOf(hash) !== -1) {
    return createHash(hash).update(input).digest('hex')
  }

  throw new Error('hash ' + hash + ' not found in your platform')

}

/**
 * Generate IV
 *
 * @param {string} key
 * @param {number} [length=16]
 * @returns {string}
 */
const iv = (key, length = 16) => {

  return key.substr(length)

}

/**
 * Normalize string
 *
 * @param input
 * @returns {*}
 */
const normalizeInput = input => {

  if (input === null || typeof input === 'undefined') {
    throw new Error('required origin')
  }

  if (typeof input === 'object') {
    input = JSON.stringify(input)
  }

  if (typeof input !== 'string') {
    input = input.toString()
  }

  return input

}

/**
 * If is JSON string then parse, else just return
 *
 * @param input
 * @returns {*}
 */
const normalizeOutput = input => {
  try {
    return JSON.parse(input)
  } catch (e) {
    return input
  }
}

/**
 * Encode string
 * @param input
 * @return {string}
 */
function encode(input) {

  input = normalizeInput(input)

  const cipher = createCipheriv(this.algorithm, this.key, this.iv, this.options)
  return cipher.update(input, 'utf8', 'hex') + cipher.final('hex')

}

/**
 * Decode string
 * @param input
 * @return {string}
 */
function decode(input) {

  input = normalizeInput(input)

  const decipher = createDecipheriv(this.algorithm, this.key, this.iv, this.options)
  const decoded = decipher.update(input, 'hex', 'utf8') + decipher.final('utf8')

  return normalizeOutput(decoded)

}


/**
 * Cryptographer
 */
export default (key, algorithm = 'aes-256-ctr', options = {}) => {

  const state = {}

  const algorithms = [
    'aes-256-cbc',
    'aes-256-cbc-hmac-sha1',
    'aes-256-cbc-hmac-sha256',
    'aes-256-cfb',
    'aes-256-cfb1',
    'aes-256-cfb8',
    'aes-256-ctr',
    'aes-256-ofb',
    'aes256',
    'camellia-256-cbc',
    'camellia-256-cfb',
    'camellia-256-cfb1',
    'camellia-256-cfb8',
    'camellia-256-ofb',
    'camellia256'
  ]

  if (typeof key !== 'string' || key === '') {
    throw new Error('required an string key')
  }


  if (algorithm !== 'aes-256-ctr' && algorithms.indexOf(algorithm) < 0) {
    throw new Error(`"${algorithm}" is not supported`)
  }


  state.algorithm = algorithm
  state.key = hash(key)
  state.iv = iv(state.key)
  state.options = options

  // console.log(state, encode.bind(state)(key))

  return {
    encode: encode.bind(state),
    decode: decode.bind(state)
  }

}


src/index.ts
import {
  createHash,
  createDecipheriv,
  createCipheriv,
  getHashes,
  CipherCCMOptions
} from 'crypto';

export declare type Algorithm = (
  'aes-256-cbc' |
  'aes-256-cbc-hmac-sha1' |
  'aes-256-cbc-hmac-sha256' |
  'aes-256-cfb' |
  'aes-256-cfb1' |
  'aes-256-cfb8' |
  'aes-256-ctr' |
  'aes-256-ofb' |
  'aes256' |
  'camellia-256-cbc' |
  'camellia-256-cfb' |
  'camellia-256-cfb1' |
  'camellia-256-cfb8' |
  'camellia-256-ofb' |
  'camellia256'
)

export declare type Data = string | object | string[] | object[] | number

const algorithms = new Set([
  'aes-256-cbc',
  'aes-256-cbc-hmac-sha1',
  'aes-256-cbc-hmac-sha256',
  'aes-256-cfb',
  'aes-256-cfb1',
  'aes-256-cfb8',
  'aes-256-ctr',
  'aes-256-ofb',
  'aes256',
  'camellia-256-cbc',
  'camellia-256-cfb',
  'camellia-256-cfb1',
  'camellia-256-cfb8',
  'camellia-256-ofb',
  'camellia256'
]);

/**
 * Creates hash of an string based on available hashes of platform
 */
const hash = (input: string, hash: string = 'md5'): string => {

  if (getHashes().indexOf(hash) !== -1) {
    return createHash(hash).update(input).digest('hex');
  }

  throw new Error('hash ' + hash + ' not found in your platform');

};

/**
 * Generate IV
 */
const iv = (key: string, length: number = 16): string => key.substring(length);

/**
 * Normalize string
 */
const normalizeInput = (input: Data): string => {

  if (input === null || typeof input === 'undefined') {
    throw new Error('required origin');
  }

  if (typeof input === 'object') {
    input = JSON.stringify(input);
  }

  if (typeof input !== 'string') {
    input = input.toString();
  }

  return input;

};

/**
 * If is JSON string then parse, else just return
 */
const normalizeOutput = (input: string): unknown => {

  try {

    return JSON.parse(input);

  } catch (e) {

    return input;

  }

};

/**
 * Encode string
 */
function encode (input: Data) {

  input = normalizeInput(input);

  const cipher = createCipheriv(
    this.algorithm,
    this.key,
    this.iv,
    this.options
  );

  return cipher.update(input, 'utf8', 'hex') + cipher.final('hex');
}

/**
 * Decode string
 */
function decode (input: any) {

  input = normalizeInput(input);

  const decipher = createDecipheriv(
    this.algorithm,
    this.key,
    this.iv,
    this.options
  );

  const decoded = decipher.update(input, 'hex', 'utf8') + decipher.final('utf8');

  return normalizeOutput(decoded);
}

/**
 * Create md5 hash string
 */
export function md5 (value: string) {

  if (typeof value !== 'string') value = String(value);

  return createHash('md5')
    .update(value)
    .digest('hex')
    .toString();
}

/**
 * Cryptographer
 */
export function Cryptographer (
  secret: string,
  algorithm?: Algorithm,
  options?: CipherCCMOptions
) {

  if (typeof algorithm === 'undefined') algorithm = 'aes-256-ctr';

  const state: Partial<{
    key: string,
    iv: string,
    algorithm: Algorithm,
    options: CipherCCMOptions
  }> = {};

  if (typeof secret !== 'string' || secret === '') {
    throw new Error('required an string key');
  }

  if (algorithm !== 'aes-256-ctr' && algorithms.has(algorithm)) {
    throw new Error(`"${algorithm}" is not supported`);
  }

  state.algorithm = algorithm;
  state.key = hash(secret);
  state.iv = iv(state.key);
  state.options = options;

  return {
    encode: encode.bind(state),
    decode: decode.bind(state)
  };

};


rollup.config.mjs
import { rollup, plugin, env, config } from '@brixtol/rollup-config';
import typescript from 'typescript';
import * as tslib from 'tslib';

export default rollup(
  {
    input: 'src/index.ts',
    output: [
      {
        format: 'es',
        dir: 'package',
        sourcemap: env.is('dev', 'inline'),
        preferConst: true
      },
      {
        format: 'cjs',
        file: config.output.main,
        sourcemap: env.is('dev', 'inline'),
        exports: 'auto'
      }
    ],
    external: [ 'crypto' ],
    plugins: env.if('dev')(
      [
        plugin.ts(
          {
            typescript,
            tslib,
            outputToFilesystem: false,
            tsconfig: 'tsconfig.json'
          }
        )
      ]
    )(
      [
        plugin.terser({ compress: { passes: 2 } })
      ]
    )
  }
);


src/keychain.js
import { createHash, createDecipheriv, createCipheriv, getCiphers, getHashes } from 'crypto'
import os from 'os'

/**
 * Creates hash of an string based on available hashes of platform
 * @param {string} input
 * @param {string} hash
 * @returns {string}
 */
const hash = (input, hash = 'md5') => {

  if (getHashes().indexOf(hash) !== -1) {
    return createHash(hash).update(input).digest('hex')
  }

  throw new Error('hash ' + hash + ' not found in your platform')

}

/**
 * Generate IV
 *
 * @param {string} key
 * @param {number} [length=16]
 * @returns {string}
 */
const iv = (key, length = 16) => {

  return key.substr(length)

}

/**
 * Normalize string
 *
 * @param input
 * @returns {*}
 */
const normalizeInput = input => {

  if (input === null || typeof input === 'undefined') throw new Error('required origin')
  if (typeof input === 'object') input = JSON.stringify(input)
  if (typeof input !== 'string') input = input.toString()

  return input
}

/**
 * If is JSON string then parse
 *
 * @param input
 * @returns {*}
 */
const normalizeOutput = input => {
  try {
    return JSON.parse(input)
  } catch (e) {
    return input
  }
}

/**
 * Encode string
 * @param input
 * @return {string}
 */
function encode (input) {

  input = normalizeInput(input)

  const cipher = createCipheriv(this.algorithm, this.key, this.iv, this.options)
  return cipher.update(input, 'utf8', 'hex') + cipher.final('hex')

}

/**
 * Decode string
 * @param input
 * @return {string}
 */
function decode (input) {

  input = normalizeInput(input)

  const decipher = createDecipheriv(this.algorithm, this.key, this.iv, this.options)
  const decoded = decipher.update(input, 'hex', 'utf8') + decipher.final('utf8')

  return normalizeOutput(decoded)

}

/**
 * State Config - provides model of each cipher
 */
const state = (

  function (keychain) {

    /**
     * Supported Encryption algorithms
     */
    const algorithms = [
      'aes-256-cbc',
      'aes-256-cbc-hmac-sha1',
      'aes-256-cbc-hmac-sha256',
      'aes-256-cfb',
      'aes-256-cfb1',
      'aes-256-cfb8',
      'aes-256-ctr',
      'aes-256-ofb',
      'aes256',
      'camellia-256-cbc',
      'camellia-256-cfb',
      'camellia-256-cfb1',
      'camellia-256-cfb8',
      'camellia-256-ofb',
      'camellia256'
    ]

    /* -------------------------------------------- */
    /*                    PRIVATE                   */
    /* -------------------------------------------- */

    /**
     * Secret
     *
     * @type {object}
     */
    let secret

    /**
     * Master
     *
     * @type {string}
     */
    let master

    /**
     * Algorithm
     *
     * @type {string}
     */
    let algorithm = 'aes-256-ctr'

    /* -------------------------------------------- */
    /*                    CLOSURE                   */
    /* -------------------------------------------- */

    return {

      /**
       * Get Master key - Returns the master key encode
       *
       * @returns {string}
       */
      get master () {

        return master

      },

      /**
       * Master key hash and IV setter, master keys
       * remains in cache, setter for master assigns
       * 1 time, on initialisation.
       *
       * @param {string} key The master password
       */
      set master (key) {

        master = hash(key)

      },

      /**
       * Keychain model - Saves the keychain
       *
       * - Left value is encoded secret key
       * - Right value is encoded value
       *
       * @example ['password', 'name']
       */
      get keychain () {

        // @ts-ignore
        return keychain

      },

      /**
       * Keychain model - Saves the keychain
       *
       * - Left value is encoded secret key
       * - Right value is encoded value
       *
       * @example ['password', 'name']
       * @param {string} value
       */
      set keychain (value) {

        secret = hash(value)

        if (!keychain.has(secret)) {
          keychain.set(secret, {
            iv: this.iv,
            key: encode.bind(this)(secret),
            secret: ''
          })
        }

      },

      /**
       * Key hash and IV setter
       *
       * - Left value is encoded variation name
       * - Right value is encoded secret key
       *
       * @example ['standard', 'password']
       */
      set key (value) {

        secret = hash(value)

      },

      /**
       * Key hash getter - MD5 by default, (see `hash()`)
       *
       * @returns {String}
       */
      get key () {

        return secret

      },

      /**
       * IV hash getter - Partial slice from `key`
       *
       * @example '123456789' > '6789'
       * @returns {String}
       */
      get iv () {

        return iv(this.key)

      },

      /**
       * Set Encryption algorithm
       *
       * @type {string}
       */
      set algorithm (value) {

        if (value !== algorithm && algorithms.indexOf(algorithm) < 0) {
          throw new Error(`"${algorithm}" is not supported`)
        }

        algorithm = value

      },

      /**
       * Get Encryption algorithm
       *
       * @returns {string}
       */
      get algorithm () {

        return algorithm

      },
      /**
       * Empty options object - passed to crypto
       *
       * @type {object}
       */
      options: Object.create(null)

    }
  }

)(new Map())

export const identity = () => createHash('sha256').update(os.homedir(), 'utf8').digest('hex')

/**
 * Keychain Supplier - Used for public exposed keychains
 *
 * @param {string} master The master key, unlocks all variations
 * @param {object} keys Access keys, unlocks on a per-varation basis
 */
export const keychain = (master, keys = undefined) => {

  if (!master || !keys) throw new Error('Empty keychain passed')

  if (Object.isFrozen(state)) {
    throw new Error('Keychain exists, modifications are prohibited')
  }

  state.master = master

  for (const key in keys) {
    state.keychain = key
    console.log(key, encode.bind(state)(key))
    for (const password of keys[key]) {
      state.keychain.get(state.key).secret = encode.bind(state)(password)
    }
  }

  console.log(state.keychain)
}

/**
 * Cryptographer
 */
export const secret = (key, algorithm = undefined, options = undefined) => {

  if (typeof key !== 'string' || key === '') throw new Error('required an string key')
  if (typeof algorithm === 'string') state.algorithm = algorithm
  if (typeof options === 'object') Object.assign(state.options, options)

  state.key = key

  // console.log(state, encode.bind(state)(key))

  return {
    encode: encode.bind(state),
    decode: decode.bind(state)
  }

}
